### ISSUE TITLE
Provide the ability to capture message bodies as tags in the OpenTelemetry span (activity) for the incoming message

### ISSUE THREAD

**Issue by lailabougria** [Particular/NServiceBus#6729]
#### Is your feature related to a problem? Please describe.

Currently, spans (or in .NET, activities), are created when a message is processed from the queue, but the message payload is not captured. Having the message payload available could provide additional insight into the scenario without requiring the user to have to add those tags manually for each message type. 

This feature was supported by the [community package for OpenTelemetry](https://github.com/jbogard/NServiceBus.Extensions.Diagnostics).

#### Describe the requested feature

Allow an opt-in API to capture message bodies in OpenTelemetry spans.

Challenges to consider are:
- Encrypted properties
- Databus properties
        
#### Describe alternatives you've considered

The only alternative for the user, is to manually add tags for every message type.

**Comment by bbrandt** [comment]
Please do not implement this as serializing the entire message body.  Allow users to define a class that describes what should be captured from an incoming or outgoing message body, using something like the [FluentValidation](https://docs.fluentvalidation.net/en/latest/index.html) syntax for creating a validator:
```csharp

public class CaptureMyMessageTypeProperties: CaptureProperties<MyMessageType>
{
  public CaptureMyMessageTypeProperties()
  {
    For(message => message .Foo);
    For(message => message .Bar);
  }
}
```
It'd be really nice if this was able to then create tags with property name as key, with maybe the option to override tag name to handle the occasional collision:
Foo: "value of Foo"
Bar: "value of Bar"

**Comment by lailabougria** [comment]
@bbrandt, Can you share more details on what limitations you see? Conversely, which use cases do you have to define a class to define what needs to be captured? The approach with a dedicated class would require you to define a class for every message type, which sounds like a lot of work to set up and maintain (every time the contract changes, you'd have to change this, too), when in the majority of the use cases you just want all of it.

I can see a case for not adding the message body as a single tag in the sense of data protection/reduction. Part of that may be handled by using encrypted properties (those properties would appear encrypted in the observability backend). That said, if every property on the message body appeared as a tag instead of one tag containing the serialized body, that would allow some reduction of that data using [OpenTelemetry processors](https://opentelemetry.io/docs/collector/transforming-telemetry/#adding-or-deleting-attributes).

**Comment by bbrandt** [comment]
Thanks @lailabougria.  I have captured a few details of how our team queries logs in app insights in #7284.  We primarily use properties captured from the message body as locators for a certain context.  We have our own auditing of the entire message bodies that we send to Snowflake, so we are concerned about unnecessary cost increase with sending the entire serialized message body to application insights.  

An example use case for us could be our operations team saying that they see there are holds on several orders.  One of the order ids is 1234.  So, we would query application insights using KQL for the order id, filtering to a certain message type, like `OrderProcessed` from a specific service to see any logs or exceptions.

So, we use the message properties as opt-in, if a team needs them to support a certain microservice, then they specify what properties from messages they'd like to use.  We have discussed that for certain services, like one related to orders, for any message coming through that service that has an OrderId it should be captured, so if there are 25 different message types, it'd be nice to not have to opt in to receiving the one property for all 25 message types.

We may not be the most common use case and I know there are many great API designers there, so don't treat my rough `CaptureMyMessageTypeProperties` code as anything more than demonstrating a concept.  There may be a more modern approach with C# source generators that could be much more efficient for filtering the properties.

The main idea is:
* We would like to capture some properties of the message body and attach them to the Open Telemetry span produced by an NSB handler.
* We'd like some way to filter the properties that are captured for the messages.  (If the way to specify the filtering is not provided, but there is some extension point present that is fine as well.)
* We'd like it to not be slow.

**Comment by ramonsmits** [comment]
We consider tags not to be the right storage for potentially large bodies. Bodies can also be binary so we would likely Base64 encode these too making it even larger.

We believe that IF any logging of payloads is done, it needs to be done with logging. If logging is configured for OTel that means all logs can be correlated to a trace.

Core already logs OUTGOING messages when log level is set to DEBUG. We currently do not log incoming payloads.


This already is a security risk. We do not think that payloads should be logged in a production environment. If a user really wants to they can by registering a behavior that:

- Logs the payload via `ILogger<T>`
- Adds a tag via `Activity.Current.AddTag(..)`


The idea by @bbrandt is for sure the better approach. Uplift specific message properties. We think the current APIs are sufficient and not need a specific API to uplift message properties, as multiple behaviors or handlers can be registered that can cherry pick specific properties. Via a behavior these can also be stored via a Logger scope so that all logs contain these.


@bbrandt Based on that we would like to close this issue but we wanted your thoughts on this to maybe see if a specific API does add value.

**Comment by bbrandt** [comment]
Yes. I agree that we can close this issue.  We have already been able to add behaviors to specify which message properties we'd like to log or add to the span with NSB+our own custom code.
